src/finch/autoschedule/compiler.py

```diff
@@ -136,52 +84,234 @@ class LogicLowerer:
-                loop_idxs = [
-                    idx.name
-                    for idx in with_subsequence(intersect(idxs_1, idxs_2), idxs_2)
-                ]
-                lhs_idxs = [idx.name for idx in idxs_2]
+                loop_idxs = with_subsequence(intersect(idxs_1, idxs_2), idxs_2)
                 (rhs, rhs_idxs) = compile_pointwise_logic(
-                    Reorder(Relabel(arg, idxs_1), idxs_2)
+                    Relabel(arg, idxs_1), list(loop_idxs), table_vars
+                )
+                # TODO: mostly the same as aggregate, used for explicit transpose
+                raise NotImplementedError
+
+            case Query(
+                Alias(_) as lhs,
+                Reformat(tns, Reorder(MapJoin(op, args), _) as reorder),
+            ):
+                assert isinstance(tns, TensorFormat)
+                fv = tns.fill_value
+                return self(
+                    Query(
+                        lhs,
+                        Reformat(
+                            tns,
+                            Aggregate(initwrite(fv), Literal(fv), reorder, ()),
+                        ),  # TODO: initwrite
+                    ),
+                    table_vars,
+                    dim_size_vars,
                 )
-                body = f":({lhs.name}[{','.join(lhs_idxs)}] = {rhs})"
-                for idx in loop_idxs:
-                    if Field(idx) in rhs_idxs:
-                        body = f":(for {idx} = _ \n {body} end)"
+
+            case Query(
+                Alias(name) as lhs,
+                Reformat(
+                    tns,
+                    Aggregate(Literal(op), Literal(init), Reorder(arg, idxs_2), idxs_1),
+                ),
+            ):
+                (rhs, rhs_idxs) = compile_pointwise_logic(arg, list(idxs_2), table_vars)
+                lhs_idxs = setdiff(idxs_2, idxs_1)
+                agg_res = ntn.Variable(name, tns)
+                table_vars[lhs] = agg_res
+                declaration = ntn.Declare(  # declare result tensor
+                    agg_res,
+                    ntn.Literal(init),
+                    ntn.Literal(op),
+                    tuple(ntn.Variable(f"{idx.name}_size", int) for idx in lhs_idxs),
+                )
+
+                body: ntn.Block | ntn.Loop = ntn.Block(
+                    (
+                        ntn.Increment(
+                            ntn.Access(
+                                agg_res,
+                                ntn.Update(ntn.Literal(op)),
+                                tuple(ntn.Variable(idx.name, int) for idx in lhs_idxs),
+                            ),
+                            rhs,
+                        ),
+                    )
+                )
+                for idx in idxs_2:
+                    if idx in rhs_idxs:
+                        body = ntn.Loop(
+                            ntn.Variable(idx.name, int),
+                            ntn.Variable(f"{idx.name}_size", int),
```

Comment:
I adopted a convention `<name>_size` for extend variables, so it ends as: `loop(i, i_size, body)`.

Reply:
Could we be more precise about promoting loop indices to the correct loop index type here? We need to promote the integer indices. I think shape_type of each tensor gives the index type of each arg, but we need to promote here, same as osman's PR where we implement broadcast but here we can use index names to determine which indices go where.